### PR TITLE
Show secret versions for job scheduler wo job pods

### PR DESCRIPTION
**Pull request by satr**
image

### CODE REVIEW

File: src/components/component/secrets/azure-key-vault-secret-state-table-row.tsx [hunk 1]
```diff
@@ -1,31 +1,86 @@
...
     </Table.Row>
   );
 };
 
 AzureKeyVaultSecretStateTableRow.propTypes = {
-  secretInReplica: PropTypes.shape(AzureKeyVaultSecretStatusModelValidationMap),
+  secretInConsumer: PropTypes.shape(
+    AzureKeyVaultSecretStatusModelValidationMap
+  ),
 } as PropTypes.ValidationMap<AzureKeyVaultSecretStateTableRowProps>;
```
Comment (Pespiri) [inline src/components/component/secrets/azure-key-vault-secret-state-table-row.tsx:86]: Bring stateless functions out of render component.
Some methods can be refactored into proper JSX elements.

Suggested edit of component:
```tsx
import { Table, Typography } from '@equinor/eds-core-react';
import * as PropTypes from 'prop-types';

import { Duration } from '../../time/duration';
import {
  AzureKeyVaultSecretStatusModel,
  AzureKeyVaultSecretStatusModelValidationMap,
} from '../../../models/azure-key-vault-secret-status';
import { smallScheduledBatchName } from '../../../utils/string';

export interface AzureKeyVaultSecretStateTableRowProps {
  secret: AzureKeyVaultSecretStatusModel;
}

const ConsumerSecretName = ({
  secret,
}: {
  secret: AzureKeyVaultSecretStatusModel;
}): JSX.Element => {
  let name: string;
  if (secret.batchName?.length > 0) {
    // show only first secret-version entry for pods of this batch
    name = `batch: ${smallScheduledBatchName(secret.batchName)}`;
  } else if (secret.jobName?.length > 0) {
    name = `job: ${smallScheduledBatchName(secret.jobName)}`;
  } else if (secret.replicaName.toLowerCase() === 'new jobs') {
    name = 'New job';
  } else {
    name = `replica: ${smallScheduledBatchName(secret.replicaName)}`;
  }

  return <Typography as="span">{name}</Typography>;
};

const ConsumerSecretCreated = ({
  secret,
}: {
  secret: AzureKeyVaultSecretStatusModel;
}): JSX.Element => {
  if (secret.batchName?.length > 0) {
    return <Duration start={secret.batchCreated} end={new Date()} />;
  } else if (secret.jobName?.length > 0) {
    return <Duration start={secret.jobCreated} end={new Date()} />;
  } else if (secret.replicaName.toLowerCase() === 'new jobs') {
    return <></>;
  }

  return <Duration start={secret.replicaCreated} end={new Date()} />;
};

export const AzureKeyVaultSecretStateTableRow = ({
  secret,
}: AzureKeyVaultSecretStateTableRowProps): JSX.Element => (
  <Table.Row>
    <Table.Cell>{secret.version}</Table.Cell>
    <Table.Cell>
      <Typography as="span">
        <ConsumerSecretName secret={secret} />
      </Typography>
    </Table.Cell>
    <Table.Cell>
      <ConsumerSecretCreated secret={secret} />
    </Table.Cell>
  </Table.Row>
);

AzureKeyVaultSecretStateTableRow.propTypes = {
  secret: PropTypes.shape(AzureKeyVaultSecretStatusModelValidationMap),
} as PropTypes.ValidationMap<AzureKeyVaultSecretStateTableRowProps>;
```

File: src/components/component/secrets/secret-list-item-title-azure-key-vault-item.tsx [hunk 2]
```diff
@@ -12,6 +12,27 @@ import { RequestState } from '../../../state/state-utils/request-states';
...
 import '../style.css';
 
+function getSecretInReplicaWithUniqueBatchNames(secretInReplicaList) {
+  let batchMap = new Map<string, Map<string, boolean>>(); //batchName:version:boolean
+  // this filtering is to avoid showing all job pods for the same batch with the same secret version
+  return secretInReplicaList.filter((secretInReplica) => {
+    if (
+      secretInReplica.batchName === null ||
+      secretInReplica.batchName.length === 0
+    )
```
Comment (Pespiri) [inline src/components/component/secrets/secret-list-item-title-azure-key-vault-item.tsx:22]: null check does not capture undefined
```suggestion
    if (!(secretInReplica.batchName?.length > 0))
```

File: src/components/component/secrets/secret-list-item-title-azure-key-vault-item.tsx [hunk 3]
```diff
@@ -12,6 +12,27 @@ import { RequestState } from '../../../state/state-utils/request-states';
...
+      batchMap.set(secretInReplica.batchName, new Map<string, boolean>());
+    if (!batchMap.get(secretInReplica.batchName).has(secretInReplica.version)) {
+      batchMap
+        .get(secretInReplica.batchName)
+        .set(secretInReplica.version, true); // first version for this batch
+      return true;
+    }
+    return false;
+  });
+}
```
Comment (Pespiri) [inline src/components/component/secrets/secret-list-item-title-azure-key-vault-item.tsx:34]: function could be simplified using `reduce()`
```suggestion
function filterSecrets(
  secrets: Array<Readonly<AzureKeyVaultSecretStatusModel>>
): Array<AzureKeyVaultSecretStatusModel> {
  // avoid showing duplicate secrets for job pods with same batchName and version
  return secrets?.reduce<Array<AzureKeyVaultSecretStatusModel>>((obj, x) => {
    if (
      !(x.batchName?.length > 0) ||
      !obj.find((y) => y.batchName === x.batchName && y.version === x.version)
    ) {
      obj.push(x);
    }
    return obj;
  }, []);
}
```

File: src/components/component/secrets/secret-list-item-title-azure-key-vault-item.tsx [hunk 4]
```diff
@@ -12,6 +12,27 @@ import { RequestState } from '../../../state/state-utils/request-states';
```
Comment (Pespiri) [inline src/components/component/secrets/secret-list-item-title-azure-key-vault-item.tsx:12]: suggested edit of component:
```tsx
import { Icon, Table, Typography } from '@equinor/eds-core-react';
import { stop } from '@equinor/eds-icons';
import * as PropTypes from 'prop-types';
import { useEffect, useState } from 'react';

import { AzureKeyVaultSecretStateTableRow } from './azure-key-vault-secret-state-table-row';
import { SecretListItemTitle } from './secret-list-item-title';
import { usePollAzureKeyVaultSecretState } from './use-poll-azure-key-vault-secret-state';
import { ScrimPopup } from '../../scrim-popup';
import { AzureKeyVaultSecretStatusModel } from '../../../models/azure-key-vault-secret-status';
import { SecretModel, SecretModelValidationMap } from '../../../models/secret';
import { RequestState } from '../../../state/state-utils/request-states';

import '../style.css';

export interface SecretListItemTitleAzureKeyVaultItemProps {
  appName: string;
  envName: string;
  componentName: string;
  secret: SecretModel;
}

export const SecretListItemTitleAzureKeyVaultItem = ({
  appName,
  envName,
  componentName,
  secret,
}: SecretListItemTitleAzureKeyVaultItemProps): JSX.Element => {
  const [pollingPauseState, setPollingPauseState] = useState(false);
  const [pollSecretState] = usePollAzureKeyVaultSecretState(
    appName,
    envName,
    componentName,
    secret.resource,
    secret.id,
    pollingPauseState
  );
  const [statusTableRows, setStatusTableRows] = useState<Array<JSX.Element>>(
    []
  );

  const [visibleScrim, setVisibleScrim] = useState(false);
  useEffect(() => {
    setPollingPauseState(!visibleScrim);
  }, [visibleScrim]);

  useEffect(() => {
    if (pollSecretState.status !== RequestState.SUCCESS) {
      return;
    }

    setStatusTableRows(
      pollSecretState.data
        ?.reduce<Array<AzureKeyVaultSecretStatusModel>>((obj, x) => {
          // avoid showing duplicate secrets for job pods with same batchName and version
          if (
            !(x.batchName?.length > 0) ||
            !obj.find(
              (y) => y.batchName === x.batchName && y.version === x.version
            )
          ) {
            obj.push(x);
          }
          return obj;
        }, [])
        .map((x, i) => (
          <AzureKeyVaultSecretStateTableRow key={i} secret={x} />
        )) ?? []
    );
  }, [secret, pollSecretState]);

  return (
    <>
      <Typography
        link
        as="span"
        token={{ textDecoration: 'none' }}
        onClick={() => setVisibleScrim(!!pollSecretState?.data)}
      >
        <SecretListItemTitle secret={secret} />
      </Typography>

      <ScrimPopup
        title={`${secret.resource}: ${secret.id}`}
        open={visibleScrim}
        onClose={() => setVisibleScrim(false)}
        isDismissable
      >
        <div className="secret-item-content">
          {statusTableRows?.length > 0 ? (
            <div className="grid--table-overflow">
              <Table>
                <Table.Head>
                  <Table.Row>
                    <Table.Cell>Version</Table.Cell>
                    <Table.Cell>Consumer</Table.Cell>
                    <Table.Cell>Consumer created</Table.Cell>
                  </Table.Row>
                </Table.Head>
                <Table.Body>{statusTableRows}</Table.Body>
              </Table>
            </div>
          ) : (
            <div className="stat_empty">
              <span>
                <Icon data={stop} />
              </span>
              <Typography>No replicas use this secret</Typography>
            </div>
          )}
        </div>
      </ScrimPopup>
    </>
  );
};

SecretListItemTitleAzureKeyVaultItem.propTypes = {
  secret: PropTypes.shape(SecretModelValidationMap).isRequired,
  appName: PropTypes.string.isRequired,
  envName: PropTypes.string.isRequired,
  componentName: PropTypes.string.isRequired,
} as PropTypes.ValidationMap<SecretListItemTitleAzureKeyVaultItemProps>;
```

File: src/models/azure-key-vault-secret-status/index.ts [hunk 5]
```diff
@@ -1,13 +1,23 @@
 import * as PropTypes from 'prop-types';
 
 export interface AzureKeyVaultSecretStatusModel {
-  replicaName?: string;
+  replicaName: string;
+  replicaCreated: Date;
```
Comment (Pespiri) [inline src/models/azure-key-vault-secret-status/index.ts:5]: is `replicaCreated` always supplied by the API or is it actually optional?

File: src/models/azure-key-vault-secret-status/index.ts [hunk 6]
```diff
@@ -1,13 +1,23 @@
...
+  batchCreated?: Date;
   version?: string;
 }
 
 /* PropTypes validation map for AzureKeyVaultSecretStatusModel */
 export const AzureKeyVaultSecretStatusModelValidationMap: PropTypes.ValidationMap<AzureKeyVaultSecretStatusModel> =
   {
-    replicaName: PropTypes.string,
+    replicaName: PropTypes.string.isRequired,
+    replicaCreated: PropTypes.instanceOf(Date).isRequired,
```
Comment (Pespiri) [inline src/models/azure-key-vault-secret-status/index.ts:17]: is `replicaCreated` always supplied by the API or is it actually optional?

**Review by Pespiri — CHANGES_REQUESTED** [review]

File: src/components/component/secrets/secret-list-item-title-azure-key-vault-item.tsx (hunk 4 above)
Comment (satr) [reply src/components/component/secrets/secret-list-item-title-azure-key-vault-item.tsx:12]: Applied

File: src/models/azure-key-vault-secret-status/index.ts (hunk 5 above)
Comment (satr) [reply src/models/azure-key-vault-secret-status/index.ts:5]: Allways

File: src/models/azure-key-vault-secret-status/index.ts (hunk 6 above)
Comment (satr) [reply src/models/azure-key-vault-secret-status/index.ts:17]: Always

File: src/components/component/secrets/secret-list-item-title-azure-key-vault-item.tsx (hunk 3 above)
Comment (satr) [reply src/components/component/secrets/secret-list-item-title-azure-key-vault-item.tsx]: Applied

File: src/components/component/secrets/azure-key-vault-secret-state-table-row.tsx (hunk 1 above)
Comment (satr) [reply src/components/component/secrets/azure-key-vault-secret-state-table-row.tsx]: Applied

**Review by Pespiri — APPROVED** [review]

**Pull request merged**